### PR TITLE
[New Integration] ProxySG

### LINKED ISSUE #9511: [New Integration] Blue Coat ProxySG
# Integration release checklist

This checklist is intended for integrations maintainers to ensure consistency
when creating or updating a Package, Module or Dataset for an Integration.

### All changes

- [x] Change follows the [contributing guidelines](https://github.com/elastic/integrations/blob/main/CONTRIBUTING.md)
- [x] Supported versions of the monitoring target are documented
- [ ] Supported operating systems are documented (if applicable)
- [x] Integration or [System tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/system_testing.md) exist
- [ ] Documentation exists, [useful guidelines](https://github.com/elastic/integrations/blob/main/docs/documentation_guidelines.md) to follow
- [x] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
- [x] At least a manual test with ES / Kibana / Agent has been performed.
- [x] Required Kibana version set to: 8.13

### New Package

- [x] Screenshot of the "Add Integration" page on Fleet added

### Dashboards changes

- [x] Dashboards exists
- [ ] Screenshots added or updated
- [ ] Datastream filters added to visualizations

### Log dataset changes

- [x] [Pipeline tests](https://github.com/elastic/elastic-package/blob/master/docs/howto/pipeline_testing.md) exist (if applicable)
- [ ] Generated output for at least 1 log file exists

# Notes

Supported log formats are described [here](https://techdocs.broadcom.com/us/en/symantec-security-software/web-and-network-security/edge-swg/7-3/getting-started/page-help-administration/page-help-logging/log-formats/default-formats.html).

The log field definitions are [here](https://techdocs.broadcom.com/us/en/symantec-security-software/web-and-network-security/edge-swg/7-3/getting-started/page-help-administration/page-help-logging/log-formats.html).

### PR DESCRIPTION
**Pull request by mjwolf**
## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Closes #9511

### CODE REVIEW

**Comment by mjwolf** [comment]
The README is still a work in progress, I think everything else is ready to review now

**Comment by mjwolf** [comment]
This PR (currently) only supports the default `main` access log format from proxysg. Support for other default logs formats will be added in the future, in other follow-up PRs. The full list of default access log formats is [here](https://techdocs.broadcom.com/us/en/symantec-security-software/web-and-network-security/edge-swg/7-3/getting-started/page-help-administration/page-help-logging/log-formats/default-formats.html)

File: packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      field: "_temp_"
+      ignore_failure: true
+# Type conversions
+  - convert:
+      tag: convert_client_to_server_port
+      field: "proxysg.client_to_server.uri_port"
+      type: long
+      ignore_missing: true
+
+# ECS mappings
```
Comment (taylor-swanson) [inline packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml:50]: Seems like there are more ECS mappings we can add, particularly with `client_to_server`

File: packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      if: ctx._temp_._conf == 'onprem'
+  - pipeline:
+      tag: cloud_pipeline
+      name: '{{ IngestPipeline "cloud" }}'
+      if: ctx._temp_._conf == 'cloud'
+
+  - set:
+      tag: set_timestamp
+      field: "@timestamp"
+      value: "{{{_temp_.date}}}T{{{_temp_.time}}}Z"
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml:37]: I've added a check

File: packages/proxysg/data_stream/log/agent/stream/udp.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    target: _temp_
+    fields:
+      _conf: "{{config}}"
+- add_locale: ~
+- syslog:
+    field: message
+    format: rfc5424
+{{#if tz_offset}}
+    timezone: "{{tz_offset}}"
+{{/if}}
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/agent/stream/udp.yml.hbs]: removed

File: packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
+---
+description: Pipeline for processing onprem ProxySG logs
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml:2]: I've removed the onprem and cloud pipelines completely. These are custom log formats, which we're not going to support initially. I've left in the pipeline structure to support different log formats; proxysg has many supported log formats, and we'll need to add support for more before this is GA

File: packages/proxysg/data_stream/log/agent/stream/stream.yml.hbs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  - preserve_original_event
+{{/if}}
+{{#each tags as |tag i|}}
+  - {{tag}}
+{{/each}}
+fields_under_root: true
+fields:
+    observer:
+        vendor: "Broadcom"
+        product: "ProxySG"
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/agent/stream/stream.yml.hbs]: Moved these to be set in the the default pipeline

File: packages/proxysg/data_stream/log/_dev/test/system/test-tls-config.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        R7lEURTXZOkkM6k5JWfgkATEmGzRxPkOloITXg9ag1OlAgMBAAGjUzBRMB0GA1Ud
+        DgQWBBRYUSKDHBBE9Q6fTeTqogicCxcXwDAfBgNVHSMEGDAWgBRYUSKDHBBE9Q6f
+        TeTqogicCxcXwDAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQBc
+        T8B+GpvPy9NQ700LsywRPY0L9IJCKiu6j3TP1tqqSPjAC/cg9ac+bFXuWOu7V+KJ
+        s09Q/pItq9SLX6UvnfRzTxu5lCBwwGX9cL131mTIu5SmFo7Eks+sorbiIarWDMoC
+        e+9An3GFpagW+YhOt4BdIM5lTqoeodzganDBsOUZI9aDAj2Yo5h2O7r6Wd12cb6T
+        mz8vMfB2eG8BxU20ZMfkdERWjiyXHOSBQqeqfkV8d9370gMu5RcJNcIgnbmTRdho
+        X3HJFiimZVaNjXATqmC/y2A1KXvJdamPLy3mGXkW2cFLoPCdK2OZFUHqiuc1bigA
+        qEf55SihFqErRMeURPPF
+        -----END CERTIFICATE-----
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/_dev/test/system/test-tls-config.yml:60]: yep, added the assert count here

File: packages/proxysg/img/screenshot-0.png
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/img/screenshot-0.png:1]: I've taken out the config screenshot

**Comment by pkoutsovasilis** [comment]
ty @taylor-swanson for adding the team tags, this was off my radar

File: packages/proxysg/data_stream/log/_dev/test/pipeline/test-proxy-main.log-expected.json [hunk 7]
```diff
@@ -0,0 +1,802 @@
+{
+    "expected": [
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/_dev/test/pipeline/test-proxy-main.log-expected.json]: I've removed the header lines from the test file. They are part of the real log files, but don't add anything to the tests

File: packages/proxysg/data_stream/log/agent/stream/stream.yml.hbs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+paths:
+{{#each paths as |path i|}}
+  - {{path}}
+{{/each}}
+exclude_files: ['\.gz$']
```
[the remark this answers is not in the document]
Comment (mjwolf) [reply packages/proxysg/data_stream/log/agent/stream/stream.yml.hbs]: I've switched to this

File: packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (mjwolf) [reply packages/proxysg/data_stream/log/elasticsearch/ingest_pipeline/default.yml:63]: I've more ECS mappings, specifically some more for `client_to_server` fields.

**Comment by mjwolf** [comment]
> I just realized that there's a screenshot of the dashboard, but there's no dashboard included in this PR. Are you planning on adding one in this PR or a follow-up PR?

I've added a dashboard now

**Comment by taylor-swanson** [comment]
I'm stilling seeing a lot of missed fields for ECS (working on compiling a list).

A bigger issue is that it appears the dashboard isn't working. I loaded the dashboard on my local stack and saw these errors. Dashboards should be standalone, any and all resources must be included with the integration. Likewise, all visualizations should be built using Lens, which as far as I know should be entirely self-contained.

Screenshot 2024-09-06 at 7 43 05 AM

**Review by taylor-swanson — COMMENTED** [review]
Remaining tasks/questions/comments:

- [ ] `proxysg.time_taken` -> `event.duration` (ensure value is converted to nanoseconds)
- [ ] `proxysg.server_to_client.status` -> `http.response.status_code` (if this is actually a proxy status, then we probably don't want to copy it to the ECS field. That's what I did for Squid)
- [ ] `proxysg.client_to_server.method` -> `http.request.method`
- [ ] `proxysg.client_to_server.host` -> `url.domain`
- [ ] `client.user.id` -> `client.user.name` (since the value is a short name, `.name` is a better fit) 
- [ ] Copy IPs to `related.ip` (i.e., `source.ip`, `destination.ip`, IPs in vendor fields, etc)
- [ ] Copy hosts to `related.hosts` (i.e.,`url.domain`, ideally the referrer host as well, but it's contained within a url)
- [ ] Copy user name to `related.user`
- [ ] Run `registered_domain` processor against `url.domain`
- [ ] Run `user_agent` processor against `user_agent.original`
- [ ] Run `geoip` against `source.ip` and `destination.ip`

Regarding the timestamp, I don't see any information on time zone. What is the time supposed to represent here? Is it the local time of the proxy, is it UTC, etc? If it's the local time, then we need to be able to provide a time zone configuration for the user. If we need the local time zone, we can add the `add_locale` processor to the agent configuration, but note that this the local time zone of the agent, not necessarily the time zone of the proxy itself (this is why we provide the time zone configuration option as well).

Additionally, I usually remove entries that are `-`, since these indicate empty values. This isn't critical and doesn't need to be done as part of this change. It's a bit more difficult to do as the values extracted from the message are put into a tree structure instead of a flat structure. For squid, I had extracted all values to a flat structure and then [used a small script](https://github.com/elastic/integrations/blob/main/packages/squid/data_stream/log/elasticsearch/ingest_pipeline/default.yml#L45) to remove any values that were `-`.

**Comment by mjwolf** [comment]
@taylor-swanson thanks, I've made all the suggestions now, except for removing '-'. I agree it's better to remove those though, I'll try to remove them later.

For the timezone, the time from proxysg is defined to always be GMT in the source logs, so I don't think it needs any other conversion.

I don't know what was wrong with the dashboard, but I completely rebuilt it, and verified that it works in a separate cluster, so I think it should be ok now.

**Comment by taylor-swanson** [comment]
> @taylor-swanson thanks, I've made all the suggestions now, except for removing '-'. I agree it's better to remove those though, I'll try to remove them later.

No worries, I have no issues with following up with that later.

> For the timezone, the time from proxysg is defined to always be GMT in the source logs, so I don't think it needs any other conversion.

Sounds good. Unix timestamps also follow a similar pattern and don't need extra processing.

**Review by taylor-swanson — APPROVED** [review]
LGTM!

**Pull request merged**